framework/midi/internal/midiconfiguration.cpp

```diff
@@ -79,43 +80,91 @@ const SynthesizerState& MidiConfiguration::synthesizerState() const
+{
+    std::list<std::string> changedGroups;
+    for (auto it = m_state.groups.cbegin(); it != m_state.groups.cend(); ++it) {
+        auto nit = state.groups.find(it->first);
+        if (nit == state.groups.cend()) {
+            continue;
+        }
+
+        if (it->second != nit->second) {
+            changedGroups.push_back(it->first);
+        }
+    }
+
+    Ret ret = writeState(stateFilePath(), state);
+    if (!ret) {
+        LOGE() << "failed write synthesizer state, file: " << stateFilePath();
+        return ret;
+    }
+
+    m_state = state;
+    m_synthesizerStateChanged.notify();
+    for (const std::string& gname : changedGroups) {
+        m_synthesizerStateGroupChanged[gname].notify();
+    }
+
+    return make_ret(Ret::Code::Ok);
+}
+
+async::Notification MidiConfiguration::synthesizerStateChanged() const
+{
+    return m_synthesizerStateChanged;
+}
+
+async::Notification MidiConfiguration::synthesizerStateGroupChanged(const std::string& gname) const
+{
+    return m_synthesizerStateGroupChanged[gname];
+}
+
 io::path MidiConfiguration::stateFilePath() const
 {
     return globalConfiguration()->dataPath() + "/synthesizer.xml";
 }
 
 bool MidiConfiguration::readState(const io::path& path, SynthesizerState& state) const
 {
-    QXmlStreamReader xml(io::pathToQString(path));
+    QFile file(io::pathToQString(path));
+    if (!file.open(QIODevice::ReadOnly)) {
+        LOGE() << "failed open " << path;
+        return false;
+    }
+
+    QXmlStreamReader xml(&file);
 
     while (xml.readNextStartElement()) {
         if (xml.name() == "Synthesizer") {
-            SynthesizerState::Group g;
-            g.name = xml.name().toString().toStdString();
-
             while (xml.readNextStartElement()) {
-                if (xml.name() == "val") {
-                    SynthesizerState::ValID id = static_cast<SynthesizerState::ValID>(xml.attributes().value("id").toInt());
-                    g.vals.push_back(SynthesizerState::Val(id, xml.readElementText().toStdString()));
-                } else {
-                    xml.skipCurrentElement();
+                SynthesizerState::Group g;
+                g.name = xml.name().toString().toStdString();
+
+                while (xml.readNextStartElement()) {
+                    if (xml.name() == "val") {
+                        SynthesizerState::ValID id = static_cast<SynthesizerState::ValID>(xml.attributes().value("id").toInt());
+                        g.vals.push_back(SynthesizerState::Val(id, xml.readElementText().toStdString()));
+                    } else {
+                        xml.skipCurrentElement();
+                    }
                 }
-            }
 
-            state.groups.push_back(std::move(g));
+                state.groups.insert({ g.name, std::move(g) });
+            }
         } else {
             LOGE() << "unknown format";
             return false;
         }
     }
 
-    return !xml.hasError();
+    return true;
```

Comment:
if reading failed, QXmlStreamReader::readNextStartElement will return false and while loop will be finished, and function will return true instead of false

Reply:
fixed